zebra-network/src/peer/handshake.rs

```diff
@@ -175,9 +180,29 @@ where
                 return Err(HandshakeError::UnexpectedMessage(remote_msg));
             }
 
-            // XXX here is where we would set the version to the minimum of the
-            // two versions, etc. -- actually is it possible to edit the `Codec`
-            // after using it to make a framed adapter?
+            // XXX in zcashd remote peer can only send one version message and we would disconnect here if it received a second one. Is it even possible for that to happen to us here?
+
+            if remote_version < constants::MIN_VERSION {
+                // Disconnect if peer is using an obsolete version.
+                return Err(HandshakeError::ObsoleteVersion(remote_version));
+            }
+
+            // TODO: Reject incoming connections from nodes that don't know about the current epoch.
+            // zcashd does this:
+            //  const Consensus::Params& consensusParams = chainparams.GetConsensus();
+            //  auto currentEpoch = CurrentEpoch(GetHeight(), consensusParams);
+            //  if (pfrom->nVersion < consensusParams.vUpgrades[currentEpoch].nProtocolVersion)
+
+            // Set the connection's version to the minimum of the received version or our own.
+            let negotiated_version = std::cmp::min(remote_version, constants::CURRENT_VERSION);
+
+            // Reconfigure the codec to use the negotiated version.
+            //
+            // XXX The tokio documentation says not to do this while any frames are still being processed.
+            // Since we don't know that here, another way might be to release the tcp
+            // stream from the unversioned Framed wrapper and construct a new one with a versioned codec.
+            let bare_codec = stream.codec_mut();
+            bare_codec.reconfigure_version(negotiated_version);
```

Comment:
Since we set the initial version for the codec to the minimum version (right? good to check?), I think that the only case that this could cause a problem would be one where the remote peer starts sending messages for a higher protocol version before negotiating with us that we actually supported it.  In that case the codec could potentially error instead of parsing those messages (if the frames were already queued), but the worst that would happen is that the connection would be closed.  Also, I think that because streams are driven by their consumers, and at this point in the code this function is the only possible consumer, the codec isn't going to be parsing anyways.

Also, as a style point (nit), I think it's better to avoid making a variable binding and just do `.codec_mut().reconfigure_version(...)` since the binding is only ever used once.